package_control/providers/repository_provider.py

```diff
@@ -168,60 +167,27 @@ def fetch_and_validate(self):
         if self.repo_info is not None:
             return True
 
-        try:
-            self.fetch()
-        except (DownloaderException, ProviderException) as e:
-            self.failed_sources[self.repo] = e
-            self.cache['get_packages'] = {}
-            return False
-
         def fail(message):
-            exception = ProviderException(message)
-            self.failed_sources[self.repo] = exception
-            self.cache['get_packages'] = {}
-            return
-        schema_error = 'Repository %s does not appear to be a valid repository file because ' % self.repo
-
-        if 'schema_version' not in self.repo_info:
-            error_string = '%s the "schema_version" JSON key is missing.' % schema_error
-            fail(error_string)
-            return False
+            raise ProviderException(
+                'Repository %s does not appear to be a valid repository file because %s' % (self.repo, message))
 
         try:
-            self.schema_version = self.repo_info.get('schema_version')
-            if isinstance(self.schema_version, int):
-                self.schema_version = float(self.schema_version)
-            if isinstance(self.schema_version, float):
-                self.schema_version = str(self.schema_version)
-        except (ValueError):
-            error_string = '%s the "schema_version" is not a valid number.' % schema_error
-            fail(error_string)
-            return False
-
-        if self.schema_version not in ['1.0', '1.1', '1.2', '2.0', '3.0.0']:
-            fail(text.format(
-                '''
-                %s the "schema_version" is not recognized. Must be one of: 1.0, 1.1, 1.2, 2.0 or 3.0.0.
-                ''',
-                schema_error
-            ))
-            return False
+            self.fetch()
 
-        version_parts = self.schema_version.split('.')
-        self.schema_major_version = int(version_parts[0])
+            try:
+                self.schema_version = SchemaVersion(self.repo_info['schema_version'])
+            except KeyError:
+                fail('the "schema_version" JSON key is missing.')
+            except ValueError as e:
+                fail(e)
 
-        if 'packages' not in self.repo_info:
-            error_string = '%s the "packages" JSON key is missing.' % schema_error
-            fail(error_string)
-            return False
+            if isinstance(self.repo_info['packages'], dict):
```

Comment:
What happened to the `if 'packages' not in self.repo_info:` check? If you meant to make the key optional, you must consider that case before accessing `self.repo_info['packages']` or use `dict.get`.

Reply:
Please read commit messages to learn about the answer (hint: aa05f44). It also reveals #1525 to be a result of incomplete reading and therefore a false positive.